agent/agent_worker.go

```diff
@@ -537,20 +547,36 @@ func (a *AgentWorker) AcquireAndRunJob(ctx context.Context, jobId string) error
 		var err error
 		var response *api.Response
 
-		acquiredJob, response, err = a.apiClient.AcquireJob(ctx, jobId)
+		acquiredJob, response, err = a.apiClient.AcquireJob(
+			ctx, jobId,
+			api.Header{Name: "X-Buildkite-Waiting-As-Locked", Value: "1"},
+			api.Header{Name: "X-Buildkite-Backoff-Sequence", Value: fmt.Sprintf("%d", r.AttemptCount())},
+		)
 		if err != nil {
-			// If the API returns with a 422, that means that we
-			// succesfully *tried* to acquire the job, but
-			// Buildkite rejected the finish for some reason.
-			if response != nil && response.StatusCode == 422 {
-				a.logger.Warn("Buildkite rejected the call to acquire the job (%s)", err)
-				r.Break()
-			} else {
-				a.logger.Warn("%s (%s)", err, r)
+			if response != nil {
+				switch response.StatusCode {
+				case http.StatusUnprocessableEntity:
+					// If the API returns with a 422, that means that we
+					// successfully *tried* to acquire the job, but
+					// Buildkite rejected the finish for some reason.
+					a.logger.Warn("Buildkite rejected the call to acquire the job (%s)", err)
+					r.Break()
+				case http.StatusLocked:
+					// If the API returns with a 423, the job is in the waiting state
+					a.logger.Warn("The job is waiting for a dependency (%s)", err)
+					duration, errParseDuration := time.ParseDuration(response.Header.Get("Retry-After") + "s")
+					if errParseDuration != nil {
+						duration = time.Second + time.Duration(rand.Int63n(int64(time.Second)))
+					}
```

Comment:
i was going to write a comment about using `strconv.Atoi` to avoid using the weird `+ "s"` thingo, but as it turns out, everything that did that was longer and/or jankier than what you've written here. good code! 👍

Reply:
It would be nice if `time.ParseDuration` had a cleaner way to specify a unit.